src/TTImp/ProcessData.idr

```diff
@@ -424,7 +425,7 @@ processData {vars} eopts nest env fc vis mbtot (MkImpLater dfc n_in ty_raw)
          arity <- getArity defs [] fullty
 
          -- Add the type constructor as a placeholder
-         tidx <- addDef n (newDef fc n linear vars fullty vis
+         tidx <- addDef n (newDef fc n linear vars fullty (collapseDefault def_vis)
```

Comment:
I think that collapses too early and you're now forgetting what the visibility setting was.